File: hardware_adapter/c/low_pass_filter.c
#include "low_pass_filter.h"
#include <stdio.h>
#include <math.h>

void lpf_init(low_pass_filter_t* filter, double alpha, bool is_angle, lpf_type_t type) {
    filter->type = type;
    filter->is_angle = is_angle;
    filter->initialized = false;
    
    if (type == LPF_TYPE_FIRST_ORDER) {
        if (alpha > 1.0) {
            printf("alpha must be less than 1\n");
            alpha = 1.0;
        }
        if (alpha < 0.0) {
            alpha = 0.0;
            printf("alpha must be greater than 0\n");
        }
    }
    filter->alpha = alpha;
    filter->prev_filtered_output = 0.0;
}

static double lpf_first_order_step(low_pass_filter_t* filter, double data_pt) {
    if (!filter->initialized) {
        filter->prev_filtered_output = data_pt;
        filter->initialized = true;
        return data_pt;
    }
    
    if (filter->is_angle) {
        double diff = angle_diff_rad((1.0 - filter->alpha) * filter->prev_filtered_output, 
                                     -(filter->alpha) * data_pt);
        filter->prev_filtered_output = diff;
    } else {
        filter->prev_filtered_output = (1.0 - filter->alpha) * filter->prev_filtered_output + 
                                        (filter->alpha) * data_pt;
    }
    return filter->prev_filtered_output;
}

static double lpf_other_step(low_pass_filter_t* filter, double data_pt) {
    if (!filter->initialized) {
        filter->prev_filtered_output = data_pt;
        filter->initialized = true;
        return data_pt;
    }
    
    if (filter->is_angle) {
        double diff = angle_diff_rad(data_pt, filter->prev_filtered_output);
        if (fabs(diff) > filter->alpha) {
            if (diff < 0) {
                filter->prev_filtered_output = angle_diff_rad(filter->prev_filtered_output, filter->alpha);
            } else {
                filter->prev_filtered_output = angle_diff_rad(filter->prev_filtered_output, -filter->alpha);
            }
        } else {
            filter->prev_filtered_output = data_pt;
        }
    }
    return filter->prev_filtered_output;
}

double lpf_step(low_pass_filter_t* filter, double data_pt) {
    if (filter->type == LPF_TYPE_FIRST_ORDER) {
        return lpf_first_order_step(filter, data_pt);
    } else if (filter->type == LPF_TYPE_OTHER) {
        return lpf_other_step(filter, data_pt);
    }
    return data_pt;
}

File: hardware_adapter/python/low_pass_filter.py
#!/bin/python3
from common import Utils, LPF_TYPE
#LPF: data = f(n-1)*a + (1-a)*f(n)
class Low_Pass_Filter():
    def __init__(self, alpha, is_angle, type:LPF_TYPE = LPF_TYPE.FIRST_ORDER):
        self._is_angle = is_angle
        self._type = type
        self._prev_filtered_output = None
        if(self._type == LPF_TYPE.FIRST_ORDER):
            if(alpha > 1):
                print("alpha must be less than 1")
                alpha = 1
            if(alpha < 0):
                alpha = 0
                print("alpha must be greater than 0")
        self._alpha = alpha

    def step(self, data_pt):
        if(self._type == LPF_TYPE.FIRST_ORDER):
            return self._first_order_step(data_pt=data_pt)
        elif(self._type == LPF_TYPE.OTHER):
            return self._other_step(data_pt=data_pt)

    def _other_step(self, data_pt):
        if(self._prev_filtered_output is None):
            self._prev_filtered_output = data_pt
            return data_pt
        if(self._is_angle):
            diff = Utils.angle_diff_rad(data_pt, self._prev_filtered_output)
            print("diff",diff,"alpha", self._alpha)
            if(abs(diff) > self._alpha):
                if(diff < 0):
                    self._prev_filtered_output = Utils.angle_diff_rad(self._prev_filtered_output, self._alpha)
                else:
                    self._prev_filtered_output = Utils.angle_diff_rad(self._prev_filtered_output, -self._alpha)
            else:
                self._prev_filtered_output = data_pt
        return self._prev_filtered_output

    def _first_order_step(self, data_pt):
        if(self._prev_filtered_output is None):
            self._prev_filtered_output = data_pt
            return data_pt
        if(self._is_angle):
            self._prev_filtered_output = Utils.angle_diff_rad((1-self._alpha) * self._prev_filtered_output, -(self._alpha)*data_pt)
        else:
            self._prev_filtered_output = (1-self._alpha) * self._prev_filtered_output + (self._alpha)*data_pt
        return self._prev_filtered_output

if __name__ == '__main__':
    import numpy as np
    filter = Low_Pass_Filter(alpha=0.8, is_angle=True, type = LPF_TYPE.FIRST_ORDER)
    out = filter.step(np.radians(4))
    print(np.degrees(out))
    out = filter.step(np.radians(20))
    print(np.degrees(out))
    out = filter.step(np.radians(40))
    print(np.degrees(out))
    out = filter.step(np.radians(30))
    print("30", np.degrees(out))